// @thirdweb-dev/wallets/evm/wallets/safe/dist/thirdweb-dev-wallets-evm-wallets-safe.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../../../dist/defineProperty-21d22449.cjs.dev.js');
var evm_wallets_base_dist_thirdwebDevWalletsEvmWalletsBase = require('../../../../dist/base-e4d8a8ff.cjs.dev.js');
var evm_connectors_safe_dist_thirdwebDevWalletsEvmConnectorsSafe = require('../../../connectors/safe/dist/thirdweb-dev-wallets-evm-connectors-safe.cjs.dev.js');
require('../../../../dist/classPrivateMethodGet-13694527.cjs.dev.js');
require('../../../../dist/checkPrivateRedeclaration-09e2947a.cjs.dev.js');
require('../../abstract/dist/thirdweb-dev-wallets-evm-wallets-abstract.cjs.dev.js');
require('ethers');
require('eventemitter3');
require('@thirdweb-dev/chains');
require('../../../../dist/tw-connector-7917f56a.cjs.dev.js');
require('@safe-global/safe-ethers-adapters');
require('@safe-global/safe-core-sdk');
require('@safe-global/safe-ethers-lib');

// re-export the connection args for convenience

class SafeWallet extends evm_wallets_base_dist_thirdwebDevWalletsEvmWalletsBase.AbstractClientWallet {
  get walletName() {
    return "Safe Wallet";
  }
  constructor(options) {
    super(SafeWallet.id, {
      ...options
    });
    defineProperty._defineProperty(this, "connector", void 0);
  }
  async getConnector() {
    if (!this.connector) {
      const {
        SafeConnector
      } = await Promise.resolve().then(function () { return require('../../../connectors/safe/dist/thirdweb-dev-wallets-evm-connectors-safe.cjs.dev.js'); });
      this.connector = new SafeConnector();
    }
    return this.connector;
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async updateChains(chains) {
    // no op
  }
  async connect(connectOptions) {
    // can't save params to storage because one of them is a class instance and can't be saved
    return await super.connect({
      ...connectOptions,
      saveParams: false
    });
  }
  getPersonalWallet() {
    return this.connector?.personalWallet;
  }
}
defineProperty._defineProperty(SafeWallet, "meta", {
  name: "Safe",
  iconURL: "ipfs://QmbbyxDDmmLQh8DzzeUR6X6B75bESsNUFmbdvS3ZsQ2pN1/SafeToken.svg"
});
defineProperty._defineProperty(SafeWallet, "id", "Safe");

exports.SafeSupportedChainsSet = evm_connectors_safe_dist_thirdwebDevWalletsEvmConnectorsSafe.SafeSupportedChainsSet;
exports.SafeWallet = SafeWallet;
