// @thirdweb-dev/react/dist/thirdweb-dev-react.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var wallets = require('@thirdweb-dev/wallets');
var useWalletConnect = require('./useWalletConnect-aa421b95.cjs.prod.js');
var paperWallet = require('./paperWallet-03f69da0.cjs.prod.js');
var safeWallet = require('./safeWallet-25afe320.cjs.prod.js');
var walletConnect = require('./walletConnect-a660b048.cjs.prod.js');
var useSafe = require('./useSafe-fb03ee65.cjs.prod.js');
var reactCore = require('@thirdweb-dev/react-core');
require('react');
require('react/jsx-runtime');
require('@tanstack/react-query');
require('@emotion/react');
require('@emotion/styled');
require('@radix-ui/react-popover');
require('copy-to-clipboard');
require('@radix-ui/react-icons');
require('@radix-ui/react-tabs');
require('fuse.js');
require('@radix-ui/react-dropdown-menu');
require('@thirdweb-dev/chains');
require('ethers');
require('tiny-invariant');
require('@radix-ui/colors');
require('@radix-ui/react-dialog');
require('detect-browser');
require('react-qr-code');
require('@radix-ui/react-tooltip');

const deviceWallet = () => {
  return {
    id: wallets.DeviceWallet.id,
    meta: wallets.DeviceWallet.meta,
    create: options => new wallets.DeviceWallet(options)
  };
};

exports.ConnectWallet = useWalletConnect.ConnectWallet;
exports.MediaRenderer = useWalletConnect.MediaRenderer;
exports.NetworkSelector = useWalletConnect.NetworkSelector;
exports.ThirdwebNftMedia = useWalletConnect.ThirdwebNftMedia;
exports.ThirdwebProvider = useWalletConnect.ThirdwebProvider;
exports.Web3Button = useWalletConnect.Web3Button;
exports.coinbaseWallet = useWalletConnect.coinbaseWallet;
exports.metamaskWallet = useWalletConnect.metamaskWallet;
exports.useCoinbaseWallet = useWalletConnect.useCoinbaseWallet;
exports.useInstalledWallets = useWalletConnect.useInstalledWallets;
exports.useMetamask = useWalletConnect.useMetamask;
exports.usePaperWallet = useWalletConnect.usePaperWallet;
exports.usePaperWalletUserEmail = useWalletConnect.usePaperWalletUserEmail;
exports.useResolvedMediaType = useWalletConnect.useResolvedMediaType;
exports.useWalletConnect = useWalletConnect.useWalletConnect;
exports.useWalletConnectV1 = useWalletConnect.useWalletConnectV1;
exports.walletConnectV1 = useWalletConnect.walletConnectV1;
exports.paperWallet = paperWallet.paperWallet;
exports.safeWallet = safeWallet.safeWallet;
exports.walletConnect = walletConnect.walletConnect;
exports.useSafe = useSafe.useSafe;
exports.deviceWallet = deviceWallet;
Object.keys(reactCore).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return reactCore[k]; }
  });
});
