// @thirdweb-dev/react/dist/thirdweb-dev-react.browser.esm.js
import { DeviceWallet } from '@thirdweb-dev/wallets';
export { C as ConnectWallet, M as MediaRenderer, N as NetworkSelector, i as ThirdwebNftMedia, T as ThirdwebProvider, W as Web3Button, c as coinbaseWallet, m as metamaskWallet, b as useCoinbaseWallet, u as useInstalledWallets, a as useMetamask, e as usePaperWallet, d as usePaperWalletUserEmail, h as useResolvedMediaType, f as useWalletConnect, g as useWalletConnectV1, w as walletConnectV1 } from './useWalletConnect-d99a1596.browser.esm.js';
export { paperWallet } from './paperWallet-a013b3fb.browser.esm.js';
export { safeWallet } from './safeWallet-0af10328.browser.esm.js';
export { walletConnect } from './walletConnect-3e2cac57.browser.esm.js';
export { u as useSafe } from './useSafe-30cf19d3.browser.esm.js';
export * from '@thirdweb-dev/react-core';
import 'react';
import 'react/jsx-runtime';
import '@tanstack/react-query';
import '@emotion/react';
import '@emotion/styled';
import '@radix-ui/react-popover';
import 'copy-to-clipboard';
import '@radix-ui/react-icons';
import '@radix-ui/react-tabs';
import 'fuse.js';
import '@radix-ui/react-dropdown-menu';
import '@thirdweb-dev/chains';
import 'ethers';
import 'tiny-invariant';
import '@radix-ui/colors';
import '@radix-ui/react-dialog';
import 'detect-browser';
import 'react-qr-code';
import '@radix-ui/react-tooltip';

const deviceWallet = () => {
  return {
    id: DeviceWallet.id,
    meta: DeviceWallet.meta,
    create: options => new DeviceWallet(options)
  };
};

export { deviceWallet };
