// @thirdweb-dev/wallets/evm/connectors/paper/dist/thirdweb-dev-wallets-evm-connectors-paper.esm.js
import { _ as _classPrivateFieldInitSpec, b as _classPrivateFieldGet, a as _classPrivateFieldSet } from '../../../../dist/classPrivateFieldSet-0ee02dfd.esm.js';
import { _ as _defineProperty } from '../../../../dist/defineProperty-e24c82ea.esm.js';
import { T as TWConnector } from '../../../../dist/tw-connector-d83550c7.esm.js';
import { UserStatus } from '@paperxyz/embedded-wallet-service-sdk';
import { utils } from 'ethers';
import { n as normalizeChainId } from '../../../../dist/normalizeChainId-dd5a6036.esm.js';
import '../../../../dist/checkPrivateRedeclaration-a6ec2e61.esm.js';
import 'eventemitter3';

const PaperChainMap = {
  1: "Ethereum",
  5: "Goerli",
  137: "Polygon",
  80001: "Mumbai",
  43114: "Avalanche"
};
var _paper = /*#__PURE__*/new WeakMap();
var _signer = /*#__PURE__*/new WeakMap();
class PaperWalletConnector extends TWConnector {
  constructor(options) {
    super();
    _defineProperty(this, "id", "paper-wallet");
    _defineProperty(this, "name", "Paper Wallet");
    _defineProperty(this, "ready", true);
    _defineProperty(this, "user", null);
    _classPrivateFieldInitSpec(this, _paper, {
      writable: true,
      value: void 0
    });
    _defineProperty(this, "options", void 0);
    _classPrivateFieldInitSpec(this, _signer, {
      writable: true,
      value: void 0
    });
    _defineProperty(this, "onAccountsChanged", async accounts => {
      if (accounts.length === 0) {
        await this.onDisconnect();
      } else {
        this.emit("change", {
          account: utils.getAddress(accounts[0])
        });
      }
    });
    _defineProperty(this, "onChainChanged", chainId => {
      const id = normalizeChainId(chainId);
      const unsupported = this.isChainUnsupported(id);
      this.emit("change", {
        chain: {
          id,
          unsupported
        }
      });
    });
    _defineProperty(this, "onDisconnect", async () => {
      this.emit("disconnect");
    });
    this.options = options;
  }
  getPaperSDK() {
    if (!_classPrivateFieldGet(this, _paper)) {
      _classPrivateFieldSet(this, _paper, new Promise(async (resolve, reject) => {
        try {
          if (!(this.options.chain.chainId in PaperChainMap)) {
            throw new Error("Unsupported chain id: " + this.options.chain.chainId);
          }
          const {
            PaperEmbeddedWalletSdk
          } = await import('@paperxyz/embedded-wallet-service-sdk');
          const chainName = PaperChainMap[this.options.chain.chainId];
          resolve(new PaperEmbeddedWalletSdk({
            clientId: this.options.clientId,
            chain: chainName
          }));
        } catch (err) {
          reject(err);
        }
      }));
    }
    return _classPrivateFieldGet(this, _paper);
  }
  async connect() {
    const paperSDK = await this.getPaperSDK();
    if (!paperSDK) {
      throw new Error("Paper SDK not initialized");
    }
    let user = await paperSDK.getUser();
    switch (user.status) {
      case UserStatus.LOGGED_OUT:
        {
          const authResult = await paperSDK.auth.loginWithPaperModal();
          this.user = authResult.user;
          break;
        }
      case UserStatus.LOGGED_IN_WALLET_INITIALIZED:
        {
          this.user = user;
          break;
        }
    }
    if (!this.user) {
      throw new Error("Error connecting User");
    }
    this.setupListeners();
    return this.getAddress();
  }
  async disconnect() {
    // await this.paper?.auth.logout();
    this.user = null;
  }
  async getAddress() {
    const signer = await this.getSigner();
    return signer.getAddress();
  }
  async isConnected() {
    try {
      const addr = await this.getAddress();
      return !!addr;
    } catch (e) {
      return false;
    }
  }
  async getProvider() {
    const signer = await this.getSigner();
    if (!signer.provider) {
      throw new Error("Provider not found");
    }
    return signer.provider;
  }
  async getSigner() {
    if (_classPrivateFieldGet(this, _signer)) {
      return _classPrivateFieldGet(this, _signer);
    }
    if (!this.user) {
      const paperSDK = await this.getPaperSDK();
      let user = await paperSDK.getUser();
      switch (user.status) {
        case UserStatus.LOGGED_IN_WALLET_INITIALIZED:
          {
            this.user = user;
            break;
          }
      }
    }
    const signer = await this.user?.wallet.getEthersJsSigner({
      rpcEndpoint: this.options.chain.rpc[0]
    });
    if (!signer) {
      throw new Error("Signer not found");
    }
    _classPrivateFieldSet(this, _signer, signer);
    return signer;
  }
  async isAuthorized() {
    return false;
  }
  async switchChain(chainId) {
    // check if chainId is supported or not
    if (!(chainId in PaperChainMap)) {
      throw new Error("Chain not supported");
    }
    const chainName = PaperChainMap[chainId];
    const chain = this.options.chains.find(c => c.chainId === chainId);
    if (!chain) {
      throw new Error("Chain not configured");
    }

    // update chain in wallet
    await this.user?.wallet.setChain({
      chain: chainName
    });

    // update signer
    _classPrivateFieldSet(this, _signer, await this.user?.wallet.getEthersJsSigner({
      rpcEndpoint: chain.rpc[0]
    }));
    this.emit("change", {
      chain: {
        id: chainId,
        unsupported: false
      }
    });
  }

  // private getUser() {
  //   if (!this.user) {
  //     throw new Error("User not found");
  //   }
  //   return this.user;
  // }

  async setupListeners() {
    const provider = await this.getProvider();
    if (provider.on) {
      provider.on("accountsChanged", this.onAccountsChanged);
      provider.on("chainChanged", this.onChainChanged);
      provider.on("disconnect", this.onDisconnect);
    }
  }
  updateChains(chains) {
    this.options.chains = chains;
  }
  isChainUnsupported(chainId) {
    return !(chainId in PaperChainMap);
  }
  async getEmail() {
    await this.getProvider();
    if (!this.user) {
      throw new Error("No user found, Paper Wallet is not connected");
    }
    return this.user.authDetails.email;
  }
}

export { PaperChainMap, PaperWalletConnector };
