// @thirdweb-dev/wallets/evm/dist/thirdweb-dev-wallets-evm.browser.esm.js
export { C as CredentialsStorage } from '../../dist/CredentialsStorage-9a5b0621.browser.esm.js';
export { a as AbstractClientWallet, D as DEFAULT_DAPP_META } from '../../dist/base-901ad5ac.browser.esm.js';
export { T as TWConnector, W as WagmiAdapter } from '../../dist/tw-connector-443c384d.browser.esm.js';
export { PaperWallet } from '../wallets/paper-wallet/dist/thirdweb-dev-wallets-evm-wallets-paper-wallet.browser.esm.js';
export { CoinbaseWallet } from '../wallets/coinbase-wallet/dist/thirdweb-dev-wallets-evm-wallets-coinbase-wallet.browser.esm.js';
export { DeviceWallet } from '../wallets/device-wallet/dist/thirdweb-dev-wallets-evm-wallets-device-wallet.browser.esm.js';
export { InjectedWallet } from '../wallets/injected/dist/thirdweb-dev-wallets-evm-wallets-injected.browser.esm.js';
export { MetaMaskWallet } from '../wallets/metamask/dist/thirdweb-dev-wallets-evm-wallets-metamask.browser.esm.js';
export { WalletConnect } from '../wallets/wallet-connect/dist/thirdweb-dev-wallets-evm-wallets-wallet-connect.browser.esm.js';
export { WalletConnectV1 } from '../wallets/wallet-connect-v1/dist/thirdweb-dev-wallets-evm-wallets-wallet-connect-v1.browser.esm.js';
export { SafeWallet } from '../wallets/safe/dist/thirdweb-dev-wallets-evm-wallets-safe.browser.esm.js';
export { a as assertWindowEthereum } from '../../dist/assertWindowEthereum-66ccef46.browser.esm.js';
export { A as AddChainError, a as ChainNotConfiguredError, C as Connector, P as ProviderRpcError, S as SwitchChainError, U as UserRejectedRequestError } from '../../dist/errors-e1a2656b.browser.esm.js';
export { n as normalizeChainId } from '../../dist/normalizeChainId-e4cc0175.browser.esm.js';
export { SafeSupportedChainsSet } from '../connectors/safe/dist/thirdweb-dev-wallets-evm-connectors-safe.browser.esm.js';
import '../../dist/classPrivateMethodGet-9d546189.browser.esm.js';
import '../../dist/checkPrivateRedeclaration-3aaaa21d.browser.esm.js';
import '../../dist/defineProperty-c8ecdc07.browser.esm.js';
import '../wallets/abstract/dist/thirdweb-dev-wallets-evm-wallets-abstract.browser.esm.js';
import 'ethers';
import 'eventemitter3';
import '@thirdweb-dev/chains';
import 'buffer';
import '../../dist/classPrivateFieldSet-eea39a9a.browser.esm.js';
import '@safe-global/safe-ethers-adapters';
import '@safe-global/safe-core-sdk';
import '@safe-global/safe-ethers-lib';
