// @thirdweb-dev/wallets/dist/CredentialsStorage-580f480e.cjs.dev.js
'use strict';

class CredentialsStorage {
  async getItem() {
    const credential = await navigator.credentials.get({
      password: true,
      unmediated: true
    });
    if (credential && "password" in credential) {
      return credential.password;
    }
    return null;
  }
  async setItem(_key, value) {
    const {
      address
    } = JSON.parse(value);
    const credentialData = {
      id: address,
      password: value
    };
    const credential = await navigator.credentials.create({
      password: credentialData
    });
    if (!credential) {
      throw new Error("Credential not created");
    }
    await navigator.credentials.store(credential);
  }
  async removeItem() {
    // Question: is there any way to remove a credential?
    const credential = await navigator.credentials.get({
      password: true,
      unmediated: true
    });
    if (credential) {
      await navigator.credentials.preventSilentAccess();
    }
  }
}

exports.CredentialsStorage = CredentialsStorage;
