// @thirdweb-dev/wallets/dist/thirdweb-dev-wallets.browser.esm.js
export { a as AbstractClientWallet, A as AsyncLocalStorage, D as DEFAULT_DAPP_META, c as createAsyncLocalStorage } from './base-901ad5ac.browser.esm.js';
export { C as CredentialsStorage } from './CredentialsStorage-9a5b0621.browser.esm.js';
export { A as AddChainError, a as ChainNotConfiguredError, C as Connector, P as ProviderRpcError, S as SwitchChainError, U as UserRejectedRequestError } from './errors-e1a2656b.browser.esm.js';
export { n as normalizeChainId } from './normalizeChainId-e4cc0175.browser.esm.js';
export { T as TWConnector, W as WagmiAdapter } from './tw-connector-443c384d.browser.esm.js';
export { a as assertWindowEthereum } from './assertWindowEthereum-66ccef46.browser.esm.js';
export { PaperWallet } from '../evm/wallets/paper-wallet/dist/thirdweb-dev-wallets-evm-wallets-paper-wallet.browser.esm.js';
export { CoinbaseWallet } from '../evm/wallets/coinbase-wallet/dist/thirdweb-dev-wallets-evm-wallets-coinbase-wallet.browser.esm.js';
export { DeviceWallet } from '../evm/wallets/device-wallet/dist/thirdweb-dev-wallets-evm-wallets-device-wallet.browser.esm.js';
export { InjectedWallet } from '../evm/wallets/injected/dist/thirdweb-dev-wallets-evm-wallets-injected.browser.esm.js';
export { MetaMaskWallet } from '../evm/wallets/metamask/dist/thirdweb-dev-wallets-evm-wallets-metamask.browser.esm.js';
export { WalletConnect } from '../evm/wallets/wallet-connect/dist/thirdweb-dev-wallets-evm-wallets-wallet-connect.browser.esm.js';
export { WalletConnectV1 } from '../evm/wallets/wallet-connect-v1/dist/thirdweb-dev-wallets-evm-wallets-wallet-connect-v1.browser.esm.js';
export { SafeSupportedChainsSet } from '../evm/connectors/safe/dist/thirdweb-dev-wallets-evm-connectors-safe.browser.esm.js';
export { SafeWallet } from '../evm/wallets/safe/dist/thirdweb-dev-wallets-evm-wallets-safe.browser.esm.js';
import './classPrivateMethodGet-9d546189.browser.esm.js';
import './checkPrivateRedeclaration-3aaaa21d.browser.esm.js';
import './defineProperty-c8ecdc07.browser.esm.js';
import '../evm/wallets/abstract/dist/thirdweb-dev-wallets-evm-wallets-abstract.browser.esm.js';
import 'ethers';
import 'eventemitter3';
import '@thirdweb-dev/chains';
import 'buffer';
import './classPrivateFieldSet-eea39a9a.browser.esm.js';
import '@safe-global/safe-ethers-adapters';
import '@safe-global/safe-core-sdk';
import '@safe-global/safe-ethers-lib';
